flybywiresim/a32nx
# feat(mcdu): Company Routes Implementation

Fixes #1656

## To do
- [x] Ingest company route and fill out the route and from/to in the FMS
- [x] Provide a reinforcing set of scratchpad messages during the following conditions:
    - local-api server is down
    - company route cannot be found
    - company route failed validation
- [x] Provide a way to view stored company routes
- [x] Bundle the server exec in the addon

## Optional To-Do
- [x] Migrate MCDU Server to local-api
- [x] Create config file to mcdu server
- [x] Liase with installer team to provide UI page to update the config file

## Summary of Changes

Provides users the ability to load company routes that are stored in the addon folder under resources/coroutes.

This is to be used in conjunction with the [SimBridge](https://github.com/flybywiresim/simbridge)running,
company routes can be downloaded via simbrief xml datafile

## Screenshots (if necessary)



## References

* A few a320 drivers on the discord

The init A page in reference to the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual
<img width="421" alt="Screenshot 2021-12-24 at 19 20 30" src="https://user-images.githubusercontent.com/26572019/147369971-e5733899-52ba-4702-bcf5-7182ff9b1b6b.png">




## Additional context



Discord username (if different from GitHub):

## Testing instructions

Co route loading
1. generate a simbrief flight plan and download the XML datafile.
2. run the local-api.exe in the root addon folder, this will create the resources folder structure
3. dump the xml file in the coroutes folder and rename it to whatever you want BUT less than 9 characters
4. start the sim, load your aircraft and enter the coroute name in the coroute field, the FROM/TO and route (excluding SIDs and STARs) should be populated
5. modify the route in anyway, observe Co Route field is now blank (Not implemented)

Coroute viewing
1. Enter a FROM/TO which has the ICAOs that match that of a coroute (not filename, this is regarding the origin and destination ICAO within the XML) you have in the coroutes folder.
2. Observe the ROUTE page outlining the details of your coroute, press insert 
3. (optional) if you have multiple coroutes with the same ICAOs as the entered FROM/TO in the coroute folder, observe the multiple pages in the ROUTE page,
4. (optional) if the route is especially long you may have the optional to scroll, observe this by pressing the up and down keys on the MCDU
4. Press insert, observe the from/to populated as-well as the route (Excluding SIDs and STARs, CI, CRZ FL etc. as suggested by IRL pilots, as these parameters vary greatly).

MCDU Server
**NOTE: steps 1-5 won't be needed when using the installer, if #6993 gets merged then this testing step is no longer needed**
1. open `local-server.exe`
2.  The console will print a list of printers
![img](https://i.imgur.com/iGPJzcV.png)
4. copy the name of the printer (for example: `HP ENVY 5000 series [03F136]`) that you use and set that in the properties file found `resources/properties.json` and enable the printer
![img](https://i.imgur.com/PRhRniG.png)
6. restart the server
7. navigate to `localhost:3838/interfaces/mcdu`(or whichever port the server console says) and observe mcdu web client
8. Ensure the mcdu web client performs as it did previously


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
